sdk/core/azure-core/src/main/java/com/azure/core/util/SharedExecutorService.java

```diff
@@ -155,43 +160,67 @@ public static SharedExecutorService getInstance() {
+     * another area in code already had a reference to the shared instance, it will now use the passed executor service
+     * to execute tasks.
+     * <p>
+     * If the executor service is already set, this will replace it with the new executor service. If the replaced
+     * executor service was created by this class, it will be shut down.
+     * <p>
+     * If the passed executor service is null, this will throw a {@link NullPointerException}. If the passed executor
+     * service is shutdown or terminated, this will throw an {@link IllegalStateException}.
+     *
+     * @param executorService The executor service to set as the shared instance.
+     * @return The previously set {@link ExecutorService}, or null if one wasn't set or if the one set was the
+     * {@link ExecutorService} created by this class.
+     * @throws NullPointerException If the passed executor service is null.
+     * @throws IllegalStateException If the passed executor service is shutdown or terminated.
+     */
+    public static ExecutorService setExecutorService(ExecutorService executorService) {
+        // We allow for the global executor service to be set from an external source to allow for consumers of the SDK
+        // to use their own thread management to run Azure SDK tasks. This allows for the SDKs to perform deeper
+        // integration into an environment, such as the consumer environment knowing details about capacity, allowing
+        // the custom executor service to better manage resources than our more general 10x the number of processors.
+        // Another scenario could be an executor service that creates threads with specific permissions, such as
+        // allowing Azure Core or Jackson to perform deep reflection on classes that are not normally allowed.
+        Objects.requireNonNull(executorService, "'executorService' cannot be null.");
+        if (executorService.isShutdown() || executorService.isTerminated()) {
+            throw new IllegalStateException("The passed executor service is shutdown or terminated.");
+        }
+
+        ExecutorService existing = EXECUTOR_UPDATER.getAndSet(INSTANCE, executorService);
+
+        if (existing instanceof InternalExecutorService) {
+            // Only the InternalExecutorService should be shut down when setting a new ExecutorService.
+            existing.shutdown();
+            return null;
+        }
+
+        return existing;
+    }
+
+    /**
+     * Resets the state of the {@link #getInstance()} to an uninitialized state.
+     * <p>
+     * This will shut down the executor service if it was created by this class. Otherwise, it will return the executor
+     * service that was set externally, without shutting it down.
+     *
+     * @return The executor service set as the shared instance if it wasn't set by this class, otherwise null.
+     */
+    public static ExecutorService reset() {
```

Comment:
Same here - we should consider returning `void` instead and rely on getES() method for users to get the previous one before resetting.